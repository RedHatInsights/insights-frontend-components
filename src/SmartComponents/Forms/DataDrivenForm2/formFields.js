import React from 'react';
import { FormGroup, TextInput, TextArea, Select, SelectOption, Checkbox, Radio, Text, TextVariants, TextContent } from '@patternfly/react-core';
import PropTypes from 'prop-types';

import { components } from '../SchemaParsers/constants';
import { Field } from 'react-final-form';
import { composeValidators } from '../Helpers/helpers';

const selectComponent = ({ componentType, input, options, isReadOnly, isDisabled, ...rest }) => ({
    [components.TEXT_FIELD]: () => (  // eslint-disable-line react/display-name
        <TextInput
            { ...input }
            { ...rest }
            isReadOnly={ isReadOnly }
            isDisabled={ isDisabled }
        />
    ),
    [components.TEXTAREA_FIELD]: () => <TextArea { ...input } { ...rest } />, // eslint-disable-line react/display-name
    [components.SELECT_COMPONENT]: () => ( // eslint-disable-line react/display-name
        <Select { ...input } { ...rest }>
            { options.map(props => (<SelectOption key={ props.value || props.label } { ...props } label={ props.label.toString() }/>)) }
        </Select>
    ),
    [components.CHECKBOX]: () =>  <Checkbox { ...input } { ...rest }/>, // eslint-disable-line react/display-name
    [components.RADIO]: () => ( // eslint-disable-line react/display-name
        <Radio
            { ...input }
            { ...rest }
            value={ !!input.value }
            onChange={ () => input.onChange(input.value) }
        />
    )
})[componentType];

const FinalFormField = ({
    componentType,
    label,
    isRequired,
    helperText,
    meta,
    description,
    ...rest
}) => {
    const { error, touched } = meta;
    const showError = touched && error;
    return (
        <FormGroup
            isRequired={ isRequired }
            label={ label }
            fieldId={ rest.id }
            isValid={ !showError }
            helperText={ helperText }
            helperTextInvalid={ meta.error }
        >
            { description && <TextContent><Text component={ TextVariants.small }>{ description }</Text></TextContent> }
            { selectComponent({ componentType, ...rest, isValid: !showError })() }
        </FormGroup>
    );
};

FinalFormField.propTypes = {
    componentType: PropTypes.string.isRequired,
    label: PropTypes.string,
    isRequired: PropTypes.bool,
    helperText: PropTypes.string,
    meta: PropTypes.object.isRequired,
    description: PropTypes.string
};

FinalFormField.defaultProps = {
    isRequired: false,
    description: undefined
};

const FormGroupWrapper = props =>
    <Field { ...props } validate={ composeValidators(...props.validate || []) } component={ FinalFormField } />;

export const Condition = ({ when, is, children }) => {
    const shouldRender = value => Array.isArray(is) ? !!is.find(item => item === value) : value === is;
    return (
        <Field name={ when } subscription={ { value: true } }>
            { ({ input: { value }}) => shouldRender(value) ? children : null }
        </Field>
    );
};

Condition.propTypes = {
    when: PropTypes.string.isRequired,
    is: PropTypes.any,
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node
    ]).isRequired
};

const FormConditionWrapper = ({ condition, children }) => condition ? (
    <Condition { ...condition }>
        { children }
    </Condition>
) : children;

const SingleCheckboxField = ({ isReadOnly, helperText, ...props }) => (
    <Field
        { ...props }
        validate={ composeValidators(...props.validate || []) }
        type="checkbox"
        render={ rest => selectComponent({ ...rest })() }
    />
);

const MultipleChoiceList = ({ validate, ...props }) => <Field { ...props } validate={ composeValidators(...props.validate || []) }>
    { ({ label, isRequired, helperText, meta, options, ...rest }) => {
        const { error, touched } = meta;
        const showError = touched && error;
        const groupValues = rest.input.value;
        return (
            <FormGroup
                label={ label }
                isRequired={ isRequired }
                fieldId={ rest.id }
                helperText={ helperText }
                helperTextInvalid= { meta.error }
                isValid={ !showError }
            >
                { options.map(option =>
                    <Field
                        id={ `${rest.id}-${option.value}` }
                        key={ option.value }
                        { ...option }
                        name={ props.name }
                        type="checkbox"
                        render={ ({ input, meta, ...rest }) => {
                            const indexValue = groupValues.indexOf(input.value);
                            return (
                                <Checkbox
                                    aria-label={ option['aria-label'] || option.label }
                                    { ...input }
                                    { ...rest }
                                    onChange={ () => indexValue === -1
                                        ? input.onChange([ ...groupValues, input.value ])
                                        : input.onChange([ ...groupValues.slice(0, indexValue), ...groupValues.slice(indexValue + 1) ]) }
                                />
                            );} } />) }
            </FormGroup>
        );
    } }
</Field>;

const SingleChoiceList = ({ validate, ...props }) => (
    <Field
        { ...props }
        validate={ composeValidators(...props.validate || []) }
        render={ ({ input, meta, options, label, isRequired, helperText, ...rest }) => {
            const { error, touched } = meta;
            const showError = touched && error;
            return (
                <FormGroup
                    label={ label }
                    isRequired={ isRequired }
                    fieldId={ rest.id }
                    helperText={ helperText }
                    helperTextInvalid= { meta.error }
                    isValid={ !showError }
                >
                    { options.map(option => (
                        <Field
                            id={ `${rest.id}-${option.value}` }
                            key={ `${props.name}-${option.value}` }
                            name={ props.name }
                            label={ option.label }
                            value={ option.value }
                            aria-label={ option['aria-label'] || option.label }
                            type="radio"
                            componentType={ components.RADIO }
                            render={ props => selectComponent(props)() }
                        />
                    )) }
                </FormGroup>
            );} } />);

const CheckboxGroupField = ({ options, ...rest }) =>
    options ? <MultipleChoiceList options={ options } { ...rest }/>
        : (
            <SingleCheckboxField
                aria-label={ rest['aria-label'] || rest.label }
                { ...rest }
                componentType={ components.CHECKBOX }
            />
        );

const fieldMapper = type => ({
    // eslint-disable-next-line react/display-name
    [components.RADIO]: props => <SingleChoiceList { ...props }/>,
    // eslint-disable-next-line react/display-name
    [components.CHECKBOX]: props => <CheckboxGroupField { ...props }/>,
    // eslint-disable-next-line react/display-name
    [components.SELECT_COMPONENT]: props => <FormGroupWrapper componentType={ components.SELECT_COMPONENT } { ...props } />,
    // eslint-disable-next-line react/display-name
    [components.TEXTAREA_FIELD]: props => <FormGroupWrapper componentType={ components.TEXTAREA_FIELD } { ...props } />,
    // eslint-disable-next-line react/display-name
    [components.TEXT_FIELD]: props => <FormGroupWrapper componentType={ components.TEXT_FIELD } { ...props } />
})[type];

const FieldInterface = ({ meta, dataType, condition, validate, componentType, initialKey, ...props }) => (
    <FormConditionWrapper condition={ condition }>
        { fieldMapper(componentType)({
            ...props,
            validate: Array.isArray(validate) ? validate : [ validate ],
            componentType,
            id: props.id || props.name
        }) }
    </FormConditionWrapper>
);

FieldInterface.propTypes = {
    meta: PropTypes.object,
    condition: PropTypes.shape({
        when: PropTypes.string.isRequired,
        is: PropTypes.oneOfType([ PropTypes.array, PropTypes.string ]).isRequired
    }),
    validate: PropTypes.oneOfType([ PropTypes.array, PropTypes.func ]),
    componentType: PropTypes.oneOf([
        components.RADIO,
        components.CHECKBOX,
        components.SELECT_COMPONENT,
        components.TEXTAREA_FIELD,
        components.TEXT_FIELD
    ]).isRequired,
    id: PropTypes.string,
    name: PropTypes.string.isRequired,
    dataType: PropTypes.any,
    initialKey: PropTypes.any
};

export const TextField = props => <FieldInterface { ...props } componentType={ components.TEXT_FIELD }  />;

export const TextareaField = props => <FieldInterface { ...props }  componentType={ components.TEXTAREA_FIELD } />;

export const SelectField = props=> <FieldInterface { ...props } componentType={ components.SELECT_COMPONENT } />;

export const RadioGroup = props => <FieldInterface { ...props } componentType={ components.RADIO }/>;

export const CheckboxGroup = props => <FieldInterface { ...props } componentType={ components.CHECKBOX }/>;
