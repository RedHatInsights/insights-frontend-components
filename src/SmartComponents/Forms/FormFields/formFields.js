import React, { Fragment } from 'react';
import { TextInput, TextArea, Checkbox, Radio, Select, SelectOption } from '@patternfly/react-core';
import PropTypes from 'prop-types';

const selectComponent = ({ componentType, input, options, groupValues, ...rest }) => ({
    textField: () => <TextInput id={ input.name } { ...input } { ...rest } />, // eslint-disable-line react/display-name
    textarea: () => <TextArea id={ input.name } { ...input } { ...rest } />, // eslint-disable-line react/display-name
    switch: () => <Checkbox id={ input.name } { ...input } value={ !!input.value } { ...rest } />, // eslint-disable-line react/display-name
    radio: () => <Radio // eslint-disable-line react/display-name
        id={ input.name }
        { ...input }
        onChange={ () => input.onChange(input.value) }
        value={ !!input.value }
        { ...rest }
    />,
    select: () => ( // eslint-disable-line react/display-name
        <Select id={ input.name } { ...input } { ...rest } value={ input.value || 'empty' } >
            { options.map(({ label, value, disabled }) => <SelectOption key={ value } value={ value } label={ label } isDisabled={ disabled } />) }
        </Select>
    ),
    checkbox: () => { // eslint-disable-line react/display-name
        const indexValue = groupValues.indexOf(input.value);
        return <Checkbox id={ `${input.name}-${input.value}` } { ...input } value={ indexValue !== -1 } onChange={ () => {
            indexValue === -1
                ? input.onChange([ ...groupValues, input.value ])
                : input.onChange([ ...groupValues.slice(0, indexValue), ...groupValues.slice(indexValue + 1) ]);
        } } { ...rest } />;
    }
})[componentType];

const FinalFormField = ({
    componentType,
    label,
    isRequired,
    helperText,
    meta,
    ...rest
}) => {
    const { error, touched } = meta;
    const showError = touched && error;

    return (
        <div className="pf-c-form__group">
            <label className={ `pf-c-form__label ${isRequired ? 'pf-m-required' : ''}` } htmlFor={ rest.id }>{ label }</label>
            { selectComponent({ componentType,  value: rest.input.value || rest.default || '', ...rest, isValid: !showError })() }
            { !showError && helperText && <p className="pf-c-form__helper-text">{ helperText }</p> }
            { showError && <p id={ `${rest.name}-helper` } className="pf-c-form__helper-text pf-m-error">{ error }</p> }
        </div>
    );
};

FinalFormField.propTypes = {
    componentType: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    isRequired: PropTypes.bool,
    helperText: PropTypes.string,
    meta: PropTypes.object
};

FinalFormField.defaultProps = {
    isRequired: false
};

export const TextField = props => <FinalFormField componentType="textField" { ...props } />;

export const TextareaField = props => <FinalFormField componentType="textarea" { ...props } />;

export const SwitchField = props => <FinalFormField componentType="switch" { ...props } />;

export const RadioField = props => <FinalFormField componentType="radio" { ...props } />;

export const CheckboxField = props => <FinalFormField componentType="checkbox" { ...props } />;

export const SelectField = props => <FinalFormField componentType="select" { ...props } />;

TextField.propTypes = {
    type: PropTypes.string,
    label: PropTypes.string.isRequired,
    isRequired: PropTypes.bool,
    helperText: PropTypes.string,
    meta: PropTypes.object
};

TextField.defaultProps = {
    type: 'text',
    isRequired: false,
    helperText: undefined,
    meta: {}
};

export const InputGroup = ({ description, title, type, children, ...rest }) => {
    return (
        <Fragment>
            { title && <h3>{ title }</h3> }
            { description && <p>{ description }</p> }
            { children }
        </Fragment>
    );
};

InputGroup.propTypes = {
    description: PropTypes.string,
    type: PropTypes.string,
    title: PropTypes.string,
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node
    ]).isRequired
};

InputGroup.defaultProps = {
    description: undefined,
    title: undefined,
    children: undefined
};
