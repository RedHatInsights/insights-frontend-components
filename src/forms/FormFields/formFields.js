import React from 'react';
import { TextInput, TextArea, Checkbox } from '@patternfly/react-core';
import PropTypes from 'prop-types';

const selectComponent = ({ componentType, input, ...rest }) => ({
    textField: <TextInput id={ input.name } { ...input } { ...rest } />,
    textarea: <TextArea id={ input.name } { ...input } { ...rest } />,
    switch: <Checkbox id={ input.name } { ...input } value={ !!input.value } { ...rest } />
})[componentType];

const FinalFormField = ({
    componentType,
    label,
    isRequired,
    helperText,
    meta,
    ...rest
}) => {
    const { error, touched } = meta;
    const showError = touched && error;
    return (
        <div className="pf-c-form__group">
            <label className={ `pf-c-form__label ${isRequired ? 'pf-m-required' : ''}` } htmlFor={ rest.id }>{ label }</label>
            { selectComponent({ componentType, ...rest, isValid: !showError }) }
            { !showError && helperText && <p className="pf-c-form__helper-text">{ helperText }</p> }
            { showError && <p id={ `${rest.name}-helper` } className="pf-c-form__helper-text pf-m-error">{ error }</p> }
        </div>
    );
};

FinalFormField.propTypes = {
    componentType: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    isRequired: PropTypes.bool,
    helperText: PropTypes.string,
    meta: PropTypes.object
};

FinalFormField.defaultProps = {
    isRequired: false
};

export const TextField = props => <FinalFormField componentType="textField" { ...props } />;

export const TextareaField = props => <FinalFormField componentType="textarea" { ...props } />;

export const SwitchField = props => <FinalFormField componentType="switch" { ...props } />;

TextField.propTypes = {
    type: PropTypes.string,
    label: PropTypes.string.isRequired,
    isRequired: PropTypes.bool,
    helperText: PropTypes.string,
    meta: PropTypes.object
};

TextField.defaultProps = {
    type: 'text',
    isRequired: false,
    helperText: undefined,
    meta: {}
};
