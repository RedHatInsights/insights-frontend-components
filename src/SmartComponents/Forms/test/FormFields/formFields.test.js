import React from 'react';
import { TextField, InputGroup } from '../../FormFields/formFields';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('Final form fields', () => {
  describe('Input group component', () => {
    it('should render input group without title and description', () => {
      const wrapper = mount(<InputGroup><h1>Children of input group</h1></InputGroup>)
      expect(toJson(wrapper)).toMatchSnapshot()
    });

    it('should render input group with title', () => {
      const wrapper = mount(<InputGroup title="Input group title"><h1>Children of input group</h1></InputGroup>)
      expect(toJson(wrapper)).toMatchSnapshot()
    });

    it('should render input group with description', () => {
      const wrapper = mount(<InputGroup description="Input group description"><h1>Children of input group</h1></InputGroup>)
      expect(toJson(wrapper)).toMatchSnapshot()
    });

    it('should render input group with title and description', () => {
      const wrapper = mount(<InputGroup title="Input group title" description="Input group description"><h1>Children of input group</h1></InputGroup>)
      expect(toJson(wrapper)).toMatchSnapshot()
    });
  });

  describe('Text field component', () => {
    let initialProps = {}
    beforeEach(() => {
      initialProps = {
        input: {
          value: '',
          name: 'Foo',
        },
        meta: {},
        label: 'Foo',
      }
    });

    it('should render correctly', () => {
      const wrapper = mount(<TextField {...initialProps} />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should not render error message if the field was not touched', () => {
      const wrapper = mount(<TextField {...initialProps} meta={{...initialProps, error: 'Error'}} />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render error message', () => {
      const wrapper = mount(<TextField {...initialProps} meta={{...initialProps, error: 'Error', touched: true}} />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render required star on the label end', () => {
      const wrapper = mount(<TextField {...initialProps} isRequired />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render helper text', () => {
      const wrapper = mount(<TextField {...initialProps} helperText="Helper text" />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });
});