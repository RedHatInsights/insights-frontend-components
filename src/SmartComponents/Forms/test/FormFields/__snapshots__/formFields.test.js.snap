// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Final form fields Input group component should render input group with description 1`] = `
<InputGroup
  description="Input group description"
>
  <p>
    Input group description
  </p>
  <h1>
    Children of input group
  </h1>
</InputGroup>
`;

exports[`Final form fields Input group component should render input group with title 1`] = `
<InputGroup
  title="Input group title"
>
  <h3>
    Input group title
  </h3>
  <h1>
    Children of input group
  </h1>
</InputGroup>
`;

exports[`Final form fields Input group component should render input group with title and description 1`] = `
<InputGroup
  description="Input group description"
  title="Input group title"
>
  <h3>
    Input group title
  </h3>
  <p>
    Input group description
  </p>
  <h1>
    Children of input group
  </h1>
</InputGroup>
`;

exports[`Final form fields Input group component should render input group without title and description 1`] = `
<InputGroup>
  <h1>
    Children of input group
  </h1>
</InputGroup>
`;

exports[`Final form fields Text field component should not render error message if the field was not touched 1`] = `
<TextField
  input={
    Object {
      "name": "Foo",
      "value": "",
    }
  }
  isRequired={false}
  label="Foo"
  meta={
    Object {
      "error": "Error",
      "input": Object {
        "name": "Foo",
        "value": "",
      },
      "label": "Foo",
      "meta": Object {},
    }
  }
  type="text"
>
  <FinalFormField
    componentType="textField"
    input={
      Object {
        "name": "Foo",
        "value": "",
      }
    }
    isRequired={false}
    label="Foo"
    meta={
      Object {
        "error": "Error",
        "input": Object {
          "name": "Foo",
          "value": "",
        },
        "label": "Foo",
        "meta": Object {},
      }
    }
    type="text"
  >
    <div
      className="pf-c-form__group"
    >
      <label
        className="pf-c-form__label "
      >
        Foo
      </label>
      <TextInput
        aria-label={null}
        className=""
        id="Foo"
        isDisabled={false}
        isReadOnly={false}
        isRequired={false}
        isValid={true}
        name="Foo"
        onChange={[Function]}
        type="text"
        value=""
      >
        <input
          aria-invalid={false}
          aria-label={null}
          className="pf-c-form-control"
          disabled={false}
          id="Foo"
          name="Foo"
          onChange={[Function]}
          readOnly={false}
          required={false}
          type="text"
          value=""
        />
      </TextInput>
    </div>
  </FinalFormField>
</TextField>
`;

exports[`Final form fields Text field component should render correctly 1`] = `
<TextField
  input={
    Object {
      "name": "Foo",
      "value": "",
    }
  }
  isRequired={false}
  label="Foo"
  meta={Object {}}
  type="text"
>
  <FinalFormField
    componentType="textField"
    input={
      Object {
        "name": "Foo",
        "value": "",
      }
    }
    isRequired={false}
    label="Foo"
    meta={Object {}}
    type="text"
  >
    <div
      className="pf-c-form__group"
    >
      <label
        className="pf-c-form__label "
      >
        Foo
      </label>
      <TextInput
        aria-label={null}
        className=""
        id="Foo"
        isDisabled={false}
        isReadOnly={false}
        isRequired={false}
        isValid={true}
        name="Foo"
        onChange={[Function]}
        type="text"
        value=""
      >
        <input
          aria-invalid={false}
          aria-label={null}
          className="pf-c-form-control"
          disabled={false}
          id="Foo"
          name="Foo"
          onChange={[Function]}
          readOnly={false}
          required={false}
          type="text"
          value=""
        />
      </TextInput>
    </div>
  </FinalFormField>
</TextField>
`;

exports[`Final form fields Text field component should render error message 1`] = `
<TextField
  input={
    Object {
      "name": "Foo",
      "value": "",
    }
  }
  isRequired={false}
  label="Foo"
  meta={
    Object {
      "error": "Error",
      "input": Object {
        "name": "Foo",
        "value": "",
      },
      "label": "Foo",
      "meta": Object {},
      "touched": true,
    }
  }
  type="text"
>
  <FinalFormField
    componentType="textField"
    input={
      Object {
        "name": "Foo",
        "value": "",
      }
    }
    isRequired={false}
    label="Foo"
    meta={
      Object {
        "error": "Error",
        "input": Object {
          "name": "Foo",
          "value": "",
        },
        "label": "Foo",
        "meta": Object {},
        "touched": true,
      }
    }
    type="text"
  >
    <div
      className="pf-c-form__group"
    >
      <label
        className="pf-c-form__label "
      >
        Foo
      </label>
      <TextInput
        aria-label={null}
        className=""
        id="Foo"
        isDisabled={false}
        isReadOnly={false}
        isRequired={false}
        isValid={false}
        name="Foo"
        onChange={[Function]}
        type="text"
        value=""
      >
        <input
          aria-invalid={true}
          aria-label={null}
          className="pf-c-form-control"
          disabled={false}
          id="Foo"
          name="Foo"
          onChange={[Function]}
          readOnly={false}
          required={false}
          type="text"
          value=""
        />
      </TextInput>
      <p
        className="pf-c-form__helper-text pf-m-error"
        id="undefined-helper"
      >
        Error
      </p>
    </div>
  </FinalFormField>
</TextField>
`;

exports[`Final form fields Text field component should render helper text 1`] = `
<TextField
  helperText="Helper text"
  input={
    Object {
      "name": "Foo",
      "value": "",
    }
  }
  isRequired={false}
  label="Foo"
  meta={Object {}}
  type="text"
>
  <FinalFormField
    componentType="textField"
    helperText="Helper text"
    input={
      Object {
        "name": "Foo",
        "value": "",
      }
    }
    isRequired={false}
    label="Foo"
    meta={Object {}}
    type="text"
  >
    <div
      className="pf-c-form__group"
    >
      <label
        className="pf-c-form__label "
      >
        Foo
      </label>
      <TextInput
        aria-label={null}
        className=""
        id="Foo"
        isDisabled={false}
        isReadOnly={false}
        isRequired={false}
        isValid={true}
        name="Foo"
        onChange={[Function]}
        type="text"
        value=""
      >
        <input
          aria-invalid={false}
          aria-label={null}
          className="pf-c-form-control"
          disabled={false}
          id="Foo"
          name="Foo"
          onChange={[Function]}
          readOnly={false}
          required={false}
          type="text"
          value=""
        />
      </TextInput>
      <p
        className="pf-c-form__helper-text"
      >
        Helper text
      </p>
    </div>
  </FinalFormField>
</TextField>
`;

exports[`Final form fields Text field component should render required star on the label end 1`] = `
<TextField
  input={
    Object {
      "name": "Foo",
      "value": "",
    }
  }
  isRequired={true}
  label="Foo"
  meta={Object {}}
  type="text"
>
  <FinalFormField
    componentType="textField"
    input={
      Object {
        "name": "Foo",
        "value": "",
      }
    }
    isRequired={true}
    label="Foo"
    meta={Object {}}
    type="text"
  >
    <div
      className="pf-c-form__group"
    >
      <label
        className="pf-c-form__label pf-m-required"
      >
        Foo
      </label>
      <TextInput
        aria-label={null}
        className=""
        id="Foo"
        isDisabled={false}
        isReadOnly={false}
        isRequired={false}
        isValid={true}
        name="Foo"
        onChange={[Function]}
        type="text"
        value=""
      >
        <input
          aria-invalid={false}
          aria-label={null}
          className="pf-c-form-control"
          disabled={false}
          id="Foo"
          name="Foo"
          onChange={[Function]}
          readOnly={false}
          required={false}
          type="text"
          value=""
        />
      </TextInput>
    </div>
  </FinalFormField>
</TextField>
`;
